File: Day8/Problem_24.cpp

/* Data Structure :- Strings(Sliding Window with Two pointers)
Problem Title :- Longest Substring Without Repeating Characters
Problem Statement :-Given A String  Return the length of the longest substring without repeating characters.
(Time Complexity :- O(n) and Space Complexity :- O(n))
*/


#include <bits/stdc++.h>
using namespace std;

int longestSubstringWithoutRepeatingCharacter(string s, int n) {
    int left = 0;
    int longest = 0;

    unordered_map<char, int> seen;

    for (int right = 0; right < n; right++) {

        if (seen.find(s[right]) != seen.end() &&
            seen[s[right]] >= left) {

            left = seen[s[right]] + 1;
        }

        seen[s[right]] = right;

        longest = max(longest, right - left + 1);
    }

    return longest;
}

int main() {
    string s;
    cin >> s;

    cout << longestSubstringWithoutRepeatingCharacter(s, s.length()) << "\n";

    return 0;
}

File: Day8/problem_24.py

"""  Data Structure :- Strings(Sliding Window with Two pointers)
Problem Title :- Longest Substring Without Repeating Characters
Problem Statement :-Given A String  Return the length of the longest substring without repeating characters.
(Time Complexity :- O(n) and Space Complexity :- O(n))

"""


def longestsubstring_withoutrepeating_character(s, n):

    left = 0
    right = 0
    longest = 0
    last_seen = {}
    while right < n :
        ch = s[right]
        if ch in last_seen and last_seen[ch] >= left:
            left = last_seen[ch] + 1
        last_seen[ch] = right
        longest = max(longest, right - left + 1)
        right += 1
    return longest

""" def longestsubstring_withoutrepeating_character(s, n):
    longest = 0

    for i in range(n):
        hashset = [0] * 256

        for j in range(i, n):
            if hashset[ord(s[j])] == 1:
                break

            hashset[ord(s[j])] = 1
            longest = max(longest, j - i + 1)

    return longest

"""

if __name__ == "__main__":
    n = int(input().strip());
    s = input().strip()
    
    ans = longestsubstring_withoutrepeating_character(s, n)
    
    print(ans)